src/main/haskell/kore/src/Kore/AST/Kore.hs

```diff
@@ -31,171 +34,182 @@ module Kore.AST.Kore
     , UnifiedSortVariable
     , UnifiedSort
     , UnifiedPattern (..)
-    , pattern UnifiedMetaPattern
-    , pattern UnifiedObjectPattern
     , asUnifiedPattern
     , transformUnifiedPattern
     ) where
 
 import Control.DeepSeq
-       ( NFData )
+       ( NFData (..) )
+import Data.Deriving
+       ( makeLiftCompare, makeLiftEq, makeLiftShowsPrec )
 import Data.Functor.Classes
 import Data.Functor.Foldable
-import GHC.Generics
-       ( Generic )
-
-import Data.Functor.Impredicative
-       ( Rotate31 (..) )
-import Kore.AST.Common
-import Kore.AST.MetaOrObject
 
+import           Kore.AST.Common
+import           Kore.AST.MetaOrObject
+import qualified Kore.Domain.Builtin as Domain
 
 {-|'UnifiedPattern' is joining the 'Meta' and 'Object' versions of 'Pattern', to
 allow using toghether both 'Meta' and 'Object' patterns.
 -}
-newtype UnifiedPattern variable child = UnifiedPattern
-    { getUnifiedPattern :: Unified (Rotate31 Pattern variable child) }
-  deriving ( Generic )
+data UnifiedPattern domain variable child where
+    UnifiedMetaPattern
+        :: !(Pattern Meta domain variable child)
+        -> UnifiedPattern domain variable child
 
-pattern UnifiedMetaPattern :: Pattern Meta var child -> UnifiedPattern var child
-pattern UnifiedMetaPattern pat = UnifiedPattern (UnifiedMeta (Rotate31 pat))
+    UnifiedObjectPattern
+        :: !(Pattern Object domain variable child)
+        -> UnifiedPattern domain variable child
 
-pattern UnifiedObjectPattern :: Pattern Object var child -> UnifiedPattern var child
-pattern UnifiedObjectPattern pat = UnifiedPattern (UnifiedObject (Rotate31 pat))
+$(return [])  -- Begin a new definition group where UnifiedPattern is in scope.
```

Comment:
It may be obvious for the others in our team, but, for my education, could you explain what's happening here? I see the comment, but I don't get it - wouldn't UnifiedPattern be in scope below even without doing anything special?

Reply:
@virgil-serbanuta `UnifiedPattern` is in scope for expressions, but not for quoting, so that quoting the type name below (`''UnifiedPattern`) would fail. Template Haskell limits the ways in which definitions can be mutually recursive; a quoted name must be defined in a previous _definition group_ before it can be used. A top-level Template Haskell splice always starts a new definition group, even if the splice is empty, as is the case here.

Thank you for asking about this. I doubt it is clear to everyone on the team; this is an unusual use of Template Haskell. I certainly have never seen this done outside this project!

Reply:
Thank you!